### PR TITLE
fix(backend): memory leak in memory caches

### LINKED ISSUE #14310: Memory caches have an infinite lifetime and grow to consume infinite RAM
### 💡 Summary

Several memory caches do not have an eviction policy and can grow to contain millions of entries per active process. This is highly inefficient and is a partial cause of the slow memory leak that affects some instances. In particular, these caches have an infinite eviction time:
* User objects (this cache is duplicated 4 additional times with varying keys)
* Public keys (this cache is duplicated with an alternate key)
* Client app metadata

As a solution, caches should define a maximum capacity and reasonable eviction time. If the capacity is exceeded, then the oldest (least recently used) entry should be evicted to make room. This will prevent caches from growing indefinitely while still ensuring a high cache hit rate.

---

This bug was first discovered in the Sharkey fork, but exists in Misskey as well. For additional details, see [Sharkey issue #600](https://activitypub.software/TransFem-org/Sharkey/-/issues/600) and [Sharkey MR #580](https://activitypub.software/TransFem-org/Sharkey/-/merge_requests/580).

Related to #14311.

### 🥰 Expected Behavior

Caches should have a limited size and reasonable eviction method to prevent infinite memory usage.

### 🤬 Actual Behavior

Memory caches are unbounded and consume infinite server memory.

### 📝 Steps to Reproduce

Run a federated instance for an extended period of time and observe memory usage. Background memory will increase indefinitely as caches are filled with remote users, public keys, and other data.

### 💻 Frontend Environment

```markdown
N/A - this is a backend issue. The frontend environment does not matter.
```

### 🛰 Backend Environment (for server admin)

```markdown
N/A - this effects all instances running a Misskey version newer than March 2021. The backend environment is irrelevant.
```

### Do you want to address this bug yourself?

- [X] Yes, I will patch the bug myself and send a pull request

### LINKED ISSUE #14311: Memory cache eviction is inefficient and error-prone
### 💡 Summary

Cache eviction from `MemoryKVCache<T>` is currently based on a hard-coded 3 minute timer. When this timer ticks, the `gc()` method loops through all entries and removes any that have been evicted. This has a few issues:
1. The 3-minute interval adds considerable jitter to eviction times. Some caches have lifetimes as short as 10 minutes, which translates to a **30% margin of error**.
2. The full loop wastes CPU resources on larger caches, like user & public keys.
3. If the cache is not properly disposed, then it creates a memory leak and timer leak.

This can be solved by modifying the implementation to create a new timer per entry. When one of these timers ticks, it removes the associated row without any loops or checks. The timer ID can be stored within the cache entry and automatically cancelled when the entry is deleted. This has three benefits:
1. Eviction timing is much more accurate.
2. The full loop is removed.
3. If the cache is not disposed, then all entries will slowly evict until the cache is empty. Then it will be garbage collected and not produce a leak.

---

This bug was first discovered in the Sharkey fork, but exists in Misskey as well. For additional details, see [Sharkey issue #601](https://activitypub.software/TransFem-org/Sharkey/-/issues/601) and [Sharkey MR #580](https://activitypub.software/TransFem-org/Sharkey/-/merge_requests/580).

Related to #14310.

### 🥰 Expected Behavior

Cache eviction should be predictable and not consume significant system resources.

### 🤬 Actual Behavior

Cache eviction is slow and inconsistent.

### 📝 Steps to Reproduce

Run a federated instance for an extended period of time and wait for caches to fill. The "garbage collection" process will consume a notable amount of CPU time every 3 minutes or so.

### 💻 Frontend Environment

```markdown
N/A - this is a backend issue. The frontend environment does not matter.
```

### 🛰 Backend Environment (for server admin)

```markdown
N/A - this effects all instances running a Misskey version newer than March 2021. The backend environment is irrelevant.
```

### Do you want to address this bug yourself?

- [X] Yes, I will patch the bug myself and send a pull request

### PR DESCRIPTION
**Pull request by warriordog**
## What

This PR resolves #14310 and #14311 by replacing all "infinite" cache lifetimes and cleaning up `MemoryKVCache<T>`. Expiration lifetimes of all memory / redis caches have been reset to a reasonable number and normalized between similar caches.
Redis caches have a synchronization fix, and an optimization is applied to `MemoryKVCache.gc()`.

## Why

The linked issues pose a serious reliability issue that is affecting many instances, including my own. Resolving the memory leaks will improve stability, uptime, and performance.

Fix #14310
Fix #14311

## Additional info (optional)

This change is upstreamed from the Sharkey fork. Please see [Sharkey MR #580](https://activitypub.software/TransFem-org/Sharkey/-/merge_requests/580) for additional details and context.

This replaces #14314, which was an earlier attempt at a similar fix.

## Checklist
- [x] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [x] Test working in a local environment
- [ ] (If needed) Add story of storybook
- [x] (If needed) Update CHANGELOG.md
- [ ] (If possible) Add tests

### CODE REVIEW

File: packages/backend/src/core/RelayService.ts [hunk 1]
```diff
@@ -35,7 +35,7 @@ export class RelayService {
 		private createSystemUserService: CreateSystemUserService,
 		private apRendererService: ApRendererService,
 	) {
-		this.relaysCache = new MemorySingleCache<MiRelay[]>(1000 * 60 * 10);
+		this.relaysCache = new MemorySingleCache<MiRelay[]>(1000 * 60 * 10); // 10s
```
Comment (kakkokari-gtyih) [inline packages/backend/src/core/RelayService.ts]:
```suggestion
		this.relaysCache = new MemorySingleCache<MiRelay[]>(1000 * 60 * 10); // 10m
```

File: packages/backend/src/core/RoleService.ts [hunk 2]
```diff
@@ -127,10 +127,8 @@ export class RoleService implements OnApplicationShutdown, OnModuleInit {
 		private moderationLogService: ModerationLogService,
 		private fanoutTimelineService: FanoutTimelineService,
 	) {
-		//this.onMessage = this.onMessage.bind(this);
-
-		this.rolesCache = new MemorySingleCache<MiRole[]>(1000 * 60 * 60 * 1);
-		this.roleAssignmentByUserIdCache = new MemoryKVCache<MiRoleAssignment[]>(1000 * 60 * 60 * 1);
+		this.rolesCache = new MemorySingleCache<MiRole[]>(1000 * 60 * 60); // 1h
+		this.roleAssignmentByUserIdCache = new MemoryKVCache<MiRoleAssignment[]>(1000 * 60 * 5); // 1h
```
Comment (kakkokari-gtyih) [inline packages/backend/src/core/RoleService.ts]:
```suggestion
		this.roleAssignmentByUserIdCache = new MemoryKVCache<MiRoleAssignment[]>(1000 * 60 * 5); // 5m
```

File: packages/backend/src/queue/processors/DeliverProcessorService.ts [hunk 3]
```diff
@@ -45,7 +45,7 @@ export class DeliverProcessorService {
 		private queueLoggerService: QueueLoggerService,
 	) {
 		this.logger = this.queueLoggerService.logger.createSubLogger('deliver');
-		this.suspendedHostsCache = new MemorySingleCache<MiInstance[]>(1000 * 60 * 60);
+		this.suspendedHostsCache = new MemorySingleCache<MiInstance[]>(1000 * 60 * 60); // 1m
```
Comment (kakkokari-gtyih) [inline packages/backend/src/queue/processors/DeliverProcessorService.ts]:
```suggestion
		this.suspendedHostsCache = new MemorySingleCache<MiInstance[]>(1000 * 60 * 60); // 1h
```

File: packages/backend/src/server/NodeinfoServerService.ts [hunk 4]
```diff
@@ -134,7 +134,7 @@ export class NodeinfoServerService {
 			return document;
 		};
 
-		const cache = new MemorySingleCache<Awaited<ReturnType<typeof nodeinfo2>>>(1000 * 60 * 10);
+		const cache = new MemorySingleCache<Awaited<ReturnType<typeof nodeinfo2>>>(1000 * 60 * 10); // 10s
```
Comment (kakkokari-gtyih) [inline packages/backend/src/server/NodeinfoServerService.ts]:
```suggestion
		const cache = new MemorySingleCache<Awaited<ReturnType<typeof nodeinfo2>>>(1000 * 60 * 10); // 10m
```

**Comment by u1-liquid** [comment]
See also: https://github.com/MisskeyIO/misskey/pull/174 https://github.com/MisskeyIO/misskey/pull/289

File: packages/backend/src/misc/cache.ts [hunk 5]
```diff
@@ -77,14 +83,14 @@ export class RedisKVCache<T> {
 
 		// Cache MISS
 		const value = await this.fetcher(key);
-		this.set(key, value);
+		await this.set(key, value);
```
Comment (syuilo) [inline packages/backend/src/misc/cache.ts:90]: awaitする必要あるかしら？

Comment (kakkokari-gtyih) [reply packages/backend/src/misc/cache.ts:90]: > we don't technically need to, but fire-and-forget async stuff is always risky. What if the write fails? This worker would have a value in its memory cache, but the others would then re-calculate a new value.

> And without this, it's impossible for any code to require a consistent cache state. There's some ongoing work to synchronize cache logic through IPC, and some of it is built to notify with an ID and then re-fetch from redis. If this returns early, then it's conceivable that the notification will reach another worker before redis is updated. That could result in inconsistent state.

> Also, there won't be concurrent calls to `set` but there could be concurrent writes to Redis if this method is called twice in a row.

https://activitypub.software/TransFem-org/Sharkey/-/merge_requests/587#note_4597

Comment (syuilo) [reply packages/backend/src/misc/cache.ts:90]: 理由があるならコメントで残しておいた方が良いかも

Comment (warriordog) [reply packages/backend/src/misc/cache.ts:90]: @syuilo sorry, I don't quite understand. Are you asking me to document the reason with code comments?

Comment (kakkokari-gtyih) [reply packages/backend/src/misc/cache.ts:90]: > @syuilo sorry, I don't quite understand. Are you asking me to document the reason with code comments?

right

Comment (warriordog) [reply packages/backend/src/misc/cache.ts:90]: thanks for clarifying! I'll add comments.

Comment (warriordog) [reply packages/backend/src/misc/cache.ts:90]: I added comments to the `fetch` methods. Please let me know whether these make sense.

**Comment by syuilo** [comment]
🙏🏻

**Pull request merged**